dlog/plonk/src/index.rs

```diff
@@ -0,0 +1,166 @@
+
+This source file implements Plonk Protocol Index primitive.
+
+*****************************************************************************************************************/
+
+use commitment_dlog::{srs::SRS, QnrField, commitment::{CommitmentCurve, PolyComm}};
+use ff_fft::{DensePolynomial, Radix2EvaluationDomain as D};
+use plonk_circuits::constraints::ConstraintSystem;
+use oracle::poseidon::ArithmeticSpongeParams;
+use algebra::{AffineCurve, Zero, One};
+use array_init::array_init;
+use algebra::PrimeField;
+
+type Fr<G> = <G as AffineCurve>::ScalarField;
+type Fq<G> = <G as AffineCurve>::BaseField;
+
+pub enum SRSValue<'a, G : CommitmentCurve> {
+    Value(SRS<G>),
+    Ref(&'a SRS<G>)
+}
+
+impl<'a, G : CommitmentCurve> SRSValue<'a, G> {
+    pub fn get_ref(&self) -> & SRS<G> {
+        match self {
+            SRSValue::Value(x) => &x,
+            SRSValue::Ref(x) => x
+        }
+    }
+}
+
+pub enum SRSSpec <'a, G: CommitmentCurve>{
+    Use(&'a SRS<G>),
+    Generate
+}
+
+impl<'a, G: CommitmentCurve> SRSValue<'a, G> where G::BaseField : PrimeField {
+    pub fn generate(size: usize) -> SRS<G> {
+        SRS::<G>::create(size)
+    }
+
+    pub fn create<'b>(size: usize, spec : SRSSpec<'a, G>) -> SRSValue<'a, G>{
+        match spec {
+            SRSSpec::Use(x) => SRSValue::Ref(x),
+            SRSSpec::Generate => SRSValue::Value(Self::generate(size))
+        }
+    }
+}
+
+pub struct Index<'a, G: CommitmentCurve> where G::ScalarField : QnrField
+{
+    // constraints as Lagrange-based polynoms
+    pub cs: ConstraintSystem<Fr<G>>,
+
+    // polynomial commitment keys
+    pub srs: SRSValue<'a, G>,
+    
+    // maximal size of polynomial section
+    pub max_poly_size: usize,
+
+    // random oracle argument parameters
+    pub fr_sponge_params: ArithmeticSpongeParams<Fr<G>>,
+    pub fq_sponge_params: ArithmeticSpongeParams<Fq<G>>,
+}
+
+pub struct VerifierIndex<'a, G: CommitmentCurve>
+{
+    pub domain: D<Fr<G>>,          // evaluation domain
+    pub max_poly_size: usize,           // maximal size of polynomial section
+    pub srs: SRSValue<'a, G>,           // polynomial commitment keys
+
+    // index polynomial commitments
+    pub sigma_comm: [PolyComm<G>; 3],   // permutation commitment array
+    pub sid_comm:   PolyComm<G>,        // SID commitment
+    pub ql_comm:    PolyComm<G>,        // left input wire commitment
+    pub qr_comm:    PolyComm<G>,        // right input wire commitment
+    pub qo_comm:    PolyComm<G>,        // output wire commitment
+    pub qm_comm:    PolyComm<G>,        // multiplication commitment
+    pub qc_comm:    PolyComm<G>,        // constant wire commitment
+
+    // poseidon polynomial commitments
+    pub rcm_comm:   [PolyComm<G>; 3],   // round constant polynomia commitment array
+    pub fpm_comm:   PolyComm<G>,        // full/partial round indicator polynomial commitment
```

Comment:
Do we need the full/partial indicators? We are going to have 0 partial rounds. This is again just in the interest of optimizing verifier index size.

Reply:
This has been committed in f14103f2eb3d0e7e4c531f688185ebf46bcbe214